### DISCUSSION TITLE
Idea: Serializons - Save Model as BIM

### DISCUSSION THREAD

**Discussion by Rasenkantenstein** [TabularEditor/TabularEditor3#1088] [Ideas]
I am always using the ALM Toolkit for deployments of meta data against PowerBI or Azure SSAS. To compare a model, the ALM Toolkit required a Model.bim file. I also think the feature of saving a model to a folder is absolutely neccessary. Thus I have to save the Model.bim manually, close the model and reopen the folder structure.

 In TE2 there is this wonderful feature for serializons:
![image](https://user-images.githubusercontent.com/6643712/152387650-1043bb6d-efca-421c-813b-de7f7c75e28c.png)

Supporting this feature in TE3 would help to improve the deployment process.

**Comment by helgews** [comment]
I used to do this also. The only reason to hassle with model.bim was ALM 😊

What I’m doing now, is comparing my workspace db with the test (or prod) instance of analysis services.

But I admit it may not work for everyone. So no objections to your suggestion!

Edit: There is also mention of an «ALM like» comparison and stepwise/selective deployment option on the Roadmap of TE3 I think. That would simplify things!

**Comment by Rasenkantenstein** [comment]
I used to have a powershell script because TE2 can create a model.bim via command line. 
This script would also launch ALM Tollkit. But I have recently deleted it by accident. 
however this also was just a workaround.

**Comment by otykier** [comment]
Thanks for reporting. We'll make sure this feature is added to TE3 as well.

**Comment by Rasenkantenstein** [comment]
@otykier : I'm not very sharp with C#-scripts in TE3, but since the serialization of a BIM file is possible with CLI... Could I write a custom script to execute CLI TE to generate a bim-file for me? And if I'm really lazy would even open ALM toolkit for me?

**Comment by otykier** [comment]
@Rasenkantenstein 
Sure! .NET provides the [Process.Start](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.start?view=netframework-4.7.2) method that lets you run any Windows executable - including TE itself.

For example, the C# script below launches a Windows Explorer window:

```csharp
System.Diagnostics.Process.Start("explorer.exe");
```

But you can use to launch ALM Toolkit, Tabular Editor, or anything else basically. Have fun :-)

**Comment by Rasenkantenstein** [comment]
@otykier Thanks for the input: I've tried this but not really successfull (at least ALM Toolkit opens)
```
System.Diagnostics.Process.Start("\"C:/Program Files (x86)/Tabular Editor/TabularEditor.exe\" -Wait -NoNewWindow -PassThru -ArgumentList \"C:/git/hfdm/MODELS/INBOUND/database.json -B Model.bim\"");
System.Diagnostics.Process.Start("C:/Program Files/Power BI ALM Toolkit/Power BI ALM Toolkit/AlmToolkit.exe");
```

But I do get an error message that a file couldn't be found. I have no idea of syntax or functions, but if anyone has, please feel free to share :-)

**Comment by otykier** [comment]
Use the following code to start TabularEditor.exe, passing the arguments `C:/git/hfdm/MODELS/INBOUND/database.json -B Model.bim` and waiting for it to complete:

```csharp
var teProcess = System.Diagnostics.Process.Start("c:/Program Files (x86)/Tabular Editor/TabularEditor.exe", "C:/git/hfdm/MODELS/INBOUND/database.json -B Model.bim");
teProcess.WaitForExit();
```

**Comment by Rasenkantenstein** [comment]
@otykier : Thank you so much. Got it working thanks to you.

`var teProcess = System.Diagnostics.Process.Start("C:/Program Files (x86)/Tabular Editor/TabularEditor.exe", "PATH/database.json -B PATH/Model.bim");
teProcess.WaitForExit();
System.Diagnostics.Process.Start("C:/Program Files/Power BI ALM Toolkit/Power BI ALM Toolkit/AlmToolkit.exe");`

By right clicking on a model I can now build it as a Model.bim and automatically launch AlmToolkit:
![image](https://user-images.githubusercontent.com/6643712/157205369-96be7060-d09a-466a-ac26-117937607e12.png)
